src/Microsoft.Health.ControlPlane.CosmosDb/Features/Storage/StoredProcedures/HardDelete/hardDeleteIdentityProvider.js

```diff
@@ -0,0 +1,139 @@
+﻿/**
+* This stored procedure provides the following functionality:
+* - Completely delete the identity provider document.
+* 
+* @constructor
+* @param {string} id - The document id.
+* @param {string} eTag - - optional etag to match against when deleting an existing document.
+*/
+
+function hardDelete(id, eTag) {
+    const collection = getContext().getCollection();
+    const collectionLink = collection.getSelfLink();
+    const response = getContext().getResponse();
+
+    const errorMessages = {
+        IdNull: `${ErrorCodes.BadRequest}: The id is undefined or null.`,
+        RequestEntityTooLarge: `${ErrorCodes.RequestEntityTooLarge}: The request could not be completed.`,
+        NotFound: `${ErrorCodes.NotFound}: The id is not found.`,
+        LastIdentity: `${ErrorCodes.BadRequest}: Cannot delete as only one identity provider exists.`,
+        PreconditionFailed: `${ErrorCodes.PreconditionFailed}: One of the specified pre-condition is not met.`
+    };
+
+    // Validate input
+    if (!id) {
+        throw new Error(errorMessages.IdNull);
+    }
+
+    let deletedResourceIdList = new Array();
+
+    getIdentityProviderCountAndDelete(eTag);
+
+    function getIdentityProviderCountAndDelete(eTag) {
+
+        var doc;
+        // Find the resource and all of its history.
+        let query = {
+            query: "SELECT COUNT(r.id) docCount FROM ROOT r"
+        };
+
+        let isQueryAccepted = collection.queryDocuments(
+            collection.getSelfLink(),
+            query,
+            {},
+            function (err, documents, responseOptions) {
+                if (err) {
+                    throw err;
+                }
+
+                if (!documents || !documents.length) {
+                    return;
+                } else {
+                    doc = documents[0];
+                    if (doc.docCount === 0) {
+                        throw new Error(errorMessages.NotFound);
+                    }
+                    else if (doc.docCount === 1) {
+                        throw new Error(errorMessages.LastIdentity);
+                    }
+                    tryQueryAndHardDelete(eTag);
+                }
+            });
+
+        if (!isQueryAccepted) {
+            // We ran out of time.
+            throw new Error(errorMessage.requestEntityTooLarge);
+        }
+
+    }
+
+    function tryQueryAndHardDelete(eTag) {
+        // Find the resource and all of its history.
+        let query = {
+            query: "SELECT * FROM ROOT r WHERE r.id = @id",
+            parameters: [{ name: "@id", value: id }]
```

Comment:
eTag can be directly accessed so it doesn't need to be passed in?

Reply:
I am passing in for every function. Not sure if this is global,